### PR TITLE
syscall: precise error and logs

### PR DESCRIPTION
**Pull request by 0x0ece**

### CODE REVIEW

File: contrib/test/run_test_vectors.sh [hunk 1]
```diff
@@ -27,8 +27,9 @@ else
...
 fi
 
-LOG=$LOG_PATH/test_exec_syscall
-cat contrib/test/syscall-fixtures/*.list | xargs -n 1000 ./$OBJDIR/unit-test/test_exec_sol_compat --log-path $LOG
+# temp disabled
+# LOG=$LOG_PATH/test_exec_syscall
+# cat contrib/test/syscall-fixtures/*.list | xargs -n 1000 ./$OBJDIR/unit-test/test_exec_sol_compat --log-path $LOG
 
 LOG=$LOG_PATH/test_exec_precompiles
 cat contrib/test/precompile-fixtures/*.list | xargs -n 1000 ./$OBJDIR/unit-test/test_exec_sol_compat --log-path $LOG
```
Comment (topointon-jump) [inline contrib/test/run_test_vectors.sh:35]: Is this because the log errors now don't pass?

Comment (0x0ece) [reply contrib/test/run_test_vectors.sh]: Fixing soon - need to regen new fixtures

File: src/flamenco/log_collector/fd_log_collector.h [hunk 2]
```diff
@@ -400,37 +400,54 @@ fd_log_collector_program_success( fd_exec_instr_ctx_t * ctx ) {
...
 
-  /* TODO: we need to distinguish at least between:
-     - InstructionError
-     - EbpfError::SyscallError
-     - ProgramError?
-     https://github.com/anza-xyz/agave/blob/v2.0.6/program-runtime/src/invoke_context.rs#L535-L549 */
-  extern char const * fd_vm_strerror( int err );
+  extern char const * fd_vm_ebpf_strerror( int err );
+  extern char const * fd_vm_syscall_strerror( int err );
   extern char const * fd_executor_instr_strerror( int err );
```
Comment (ravyu-jump) [inline src/flamenco/log_collector/fd_log_collector.h:422]: OOC: why not declare these in `fd_vm_base.h` (or a separate header)?

File: src/flamenco/log_collector/fd_log_collector.h [hunk 3]
```diff
@@ -400,37 +400,54 @@ fd_log_collector_program_success( fd_exec_instr_ctx_t * ctx ) {
...
 
-  char custom_err[33];
-  const char * err;
+  char custom_err[33] = { 0 };
+  const char * err = custom_err;
   if( ctx->txn_ctx->custom_err != UINT_MAX ) {
     /* Max msg_sz = 32 <= 66 */
     snprintf( custom_err, sizeof(custom_err), "custom program error: 0x%x", ctx->txn_ctx->custom_err );
     err = (const char *)custom_err;
```
Comment (ravyu-jump) [inline src/flamenco/log_collector/fd_log_collector.h:428]: nit: remove

File: src/flamenco/runtime/tests/fd_exec_instr_test.c [hunk 4]
```diff
@@ -1817,15 +1829,10 @@ fd_exec_vm_syscall_test_run( fd_exec_instr_test_runner_t * runner,
...
-    FD_TEST_CUSTOM( logs_len==1, "We only support syscalls tests with at most 1 log msg" );
-    uchar const * msg; ulong msg_sz;
-    fd_log_collector_debug_get( log, 0, &msg, &msg_sz );
+  if( log->buf_sz ) {
     effects->log = FD_SCRATCH_ALLOC_APPEND(
-      l, alignof(uchar), PB_BYTES_ARRAY_T_ALLOCSIZE( msg_sz ) );
-    effects->log->size = (uint)msg_sz;
-    fd_memcpy( effects->log->bytes, msg, msg_sz );
+      l, alignof(uchar), PB_BYTES_ARRAY_T_ALLOCSIZE( log->buf_sz ) );
+    effects->log->size = (uint)fd_log_collector_debug_sprintf( log, (char *)effects->log->bytes, 0 )-1;
```
Comment (ravyu-jump) [inline src/flamenco/runtime/tests/fd_exec_instr_test.c:1835]: why -1?

File: src/flamenco/vm/syscall/fd_vm_syscall_util.c [hunk 5]
```diff
@@ -51,28 +54,17 @@ fd_vm_syscall_sol_panic( /**/            void *  _vm,
...
-  char const * file = FD_TRANSLATE_STRING( vm, file_vaddr, file_sz );
-
-  /* Max msg_sz: 37 - 6 + 20 + 20 + file_sz = 71 + file_sz <= 127 => file_sz <= 56 */
-  if( file_sz<=56 ) {
-    /* Note: when file_sz==0, file can be undefined. We use printf with "%.*s" to
-             protect against it. */
-    fd_log_collector_printf_dangerous_max_127( vm->instr_ctx,
-      "SBF program Panicked in %.*s at %lu:%lu", file_sz, file, line, column );
-  }
+  /* Disregard the precise log */
```
Comment (ravyu-jump) [inline src/flamenco/vm/syscall/fd_vm_syscall_util.c:63]: Why do we disregard here?

File: src/flamenco/runtime/fd_executor.c [hunk 6]
```diff
@@ -959,13 +958,21 @@ fd_execute_instr( fd_exec_txn_ctx_t * txn_ctx,
     } else {
       /* FIXME: this looks very different
          https://github.com/anza-xyz/agave/blob/v2.0.6/program-runtime/src/invoke_context.rs#L535-L549 */
+      if( !ctx->exec_err ) {
+        ctx->exec_err         = exec_result;
+        ctx->exec_err_kind    = FD_EXECUTOR_ERR_KIND_INSTR;
+      }
```
Comment (ravyu-jump) [inline src/flamenco/runtime/fd_executor.c:964]: This is very clean, nice :)

File: src/flamenco/runtime/fd_executor.c [hunk 7]
```diff
@@ -959,13 +958,21 @@ fd_execute_instr( fd_exec_txn_ctx_t * txn_ctx,
     } else {
       /* FIXME: this looks very different
```
Comment (ravyu-jump) [inline src/flamenco/runtime/fd_executor.c:959]: FIXME can be removed

**Review by ravyu-jump — COMMENTED** [review]
This looks great! Thanks

File: src/flamenco/runtime/tests/fd_exec_instr_test.c (hunk 4 above)
Comment (0x0ece) [reply src/flamenco/runtime/tests/fd_exec_instr_test.c:1835]: fixed (fd_log_collector_debug_sprintf used to return strlen+1, including the final \0)

File: src/flamenco/vm/syscall/fd_vm_syscall_util.c (hunk 5 above)
Comment (0x0ece) [reply src/flamenco/vm/syscall/fd_vm_syscall_util.c:63]: I'm truncating the log, sol_panic can't be used by rust, only by C sdk (who writes Solana Programs in C!?).
It's an unnecessary effort to make panic work like in agave.

File: src/flamenco/log_collector/fd_log_collector.h (hunk 2 above)
Comment (0x0ece) [reply src/flamenco/log_collector/fd_log_collector.h:422]: I don't know, will give it a try... 🤞

File: contrib/test/run_test_vectors.sh (hunk 1 above)
Comment (topointon-jump) [reply contrib/test/run_test_vectors.sh]: are you planning to do it in this PR, or in a follow-up?

File: src/flamenco/runtime/tests/fd_exec_instr_test.c [hunk 8]
```diff
@@ -1789,9 +1790,20 @@ fd_exec_vm_syscall_test_run( fd_exec_instr_test_runner_t * runner,
...
+  }
 
   /* Capture the effects */
-  effects->error = -syscall_err;
+  int exec_err = vm->instr_ctx->exec_err;
+  effects->error = 0;
+  if( syscall_err ) {
+    effects->error = exec_err <= 0 ? -vm->instr_ctx->exec_err : -1;
+    if( exec_err==0 ) {
+      FD_LOG_WARNING(( "TODO: syscall returns error, but exec_err not set. this is probably missing a log." ));
```
Comment (topointon-jump) [inline src/flamenco/runtime/tests/fd_exec_instr_test.c:1803]: wondering if this will get too spammy.. maybe it should be NOTICE? after we've fixed all the syscalls we can flip it back to WARNING.

File: src/flamenco/log_collector/fd_log_collector.h (hunk 2 above)
Comment (0x0ece) [reply src/flamenco/log_collector/fd_log_collector.h:422]: there's a circular dep. log is used by vm/exec, but this function needs them.
the core of the issue is how we defined instr_ctx, I think it's easier to address that in a separate PR.

Comment (ravyu-jump) [reply src/flamenco/log_collector/fd_log_collector.h:422]: sounds good

File: contrib/test/run_test_vectors.sh (hunk 1 above)
Comment (0x0ece) [reply contrib/test/run_test_vectors.sh]: Prob faster if we merge this to unblock other ppl, I can redo fixtures in a separated PR.

**Review by topointon-jump — APPROVED** [review]
This is awesome!

File: src/flamenco/runtime/tests/fd_exec_instr_test.c (hunk 8 above)
Comment (0x0ece) [reply src/flamenco/runtime/tests/fd_exec_instr_test.c:1803]: For the fixtures we have, we never hit that. So it should be pretty silent. This is only an edge case where we return an error but we don't track Agave's corresponding error. I've already fixed all crypto + logging, including the underlying macros like CU, memory mapping.
